crates/rattler_networking/src/s3_middleware.rs

```diff
@@ -0,0 +1,320 @@
+//! Middleware to handle `s3://` URLs to pull artifacts from an S3 bucket
+use std::path::PathBuf;
+
+use async_trait::async_trait;
+use aws_config::BehaviorVersion;
+use aws_sdk_s3::presigning::{PresignedRequest, PresigningConfig};
+use reqwest::{Request, Response};
+use reqwest_middleware::{Middleware, Next, Result as MiddlewareResult};
+use url::Url;
+
+/// S3 middleware to authenticate requests.
+pub struct S3Middleware {
+    config_file: Option<PathBuf>,
+    profile: Option<String>,
+    force_path_style: Option<bool>,
```

Comment:
services like minio don't use `https://<bucket-name>.hostname/item.txt` but `https://hostname/bucket-name/item.txt` to provide objects. by default, aws doesn't use the path style.

interestingly, awscli manages to infer this itself; will take a look how it does this soon and whether we should implement this too.

AFAIK you can't specify `use-path-style` in the aws config file :/

Reply:
This is usually specified in the AWS config file via `s3.addressing_style`, but is currently not supported by the rust AWS SDK: https://github.com/awslabs/aws-sdk-rust/issues/1230